python/cudf/cudf/core/indexing.py

```diff
@@ -266,11 +266,35 @@ def _getitem_tuple_arg(self, arg):
             else:
                 return columns_df.index._get_row_major(columns_df, arg[0])
         else:
-            df = DataFrame()
-            for col in columns_df.columns:
-                # need Series() in case a scalar is returned
-                df[col] = Series(columns_df[col].loc[arg[0]])
-            df.columns = columns_df.columns
+            if isinstance(arg[0], slice):
+                start_index, stop_index = columns_df.index.find_label_range(
+                    arg[0].start, arg[0].stop
+                )
+
+                pos_slice = slice(start_index, stop_index, arg[0].step)
+                df = columns_df[pos_slice]
+            elif is_scalar(arg[0]):
+                try:
+                    index = column.as_column(
+                        columns_df.index
+                    ).find_first_value(arg[0], closest=False)
+                except (TypeError, KeyError, IndexError, ValueError):
+                    raise IndexError("label scalar is out of bound")
+
+                df = columns_df[index : index + 1]
+            else:
+                if len(arg[0]) == 0:
+                    return columns_df[0:0]
+                arg = (column.as_column(arg[0]), arg[1])
+
+                if pd.api.types.is_bool_dtype(arg[0]):
+                    df = columns_df.take(arg[0])
+                elif isinstance(arg[0], column.ColumnBase):
+                    other_df = DataFrame()
+                    other_df.index = as_index(arg[0])
+                    df = columns_df.join(other_df, how="inner")
```

Comment:
We need to do some work to maintain the expected order here since join doesn't guarantee it.

Reply:
Would also suggest doing `other_df.join(columns_df)` as it's much more likely that `columns_df` has unique keys so we'd want that to be the right table for when joins potentially change in the future to always build the hash table against the right table.